### PR TITLE
[1 / 5] Optimize logic for gossiping assignments

### PR DESCRIPTION
**Pull request by alexggh**
This is part of the work to further optimize the approval subsystems, if you want to understand the full context start with reading https://github.com/paritytech/polkadot-sdk/pull/4849#issue-2364261568, however that's not necessary, as this change is self-contained and nodes would benefit from it regardless of subsequent changes landing or not.

While testing with 1000 validators I found out that the logic for determining the validators an assignment should be gossiped to is taking a lot of time, because it always iterated through all the peers, to determine which are X and Y neighbours and to which we should randomly gossip(4 samples).

This could be actually optimised, so we don't have to iterate through all peers for each new assignment, by fetching the list of X and Y peer ids from the topology first and then stopping the loop once we took the 4 random samples.

With this improvements we reduce the total CPU time spent in approval-distribution with 15% on networks with 500 validators and 20% on networks with 1000 validators.

## Test coverage:

`propagates_assignments_along_unshared_dimension` and `propagates_locally_generated_assignment_to_both_dimensions`  cover already logic and they passed, confirm that there is no breaking change.

Additionally, the approval voting benchmark measure the traffic sent to other peers, so I confirmed that for various network size there is no difference in the size of the traffic sent to other peers.

### CODE REVIEW

**Review by alindima — APPROVED** [review]
Nice!

**Review by ordian — APPROVED** [review]
great find!

**Review by sandreim — APPROVED** [review]
Nice work!

File: polkadot/node/network/protocol/src/grid_topology.rs [hunk 1]
```diff
@@ -524,6 +541,11 @@ impl RandomRouting {
 	pub fn inc_sent(&mut self) {
 		self.sent += 1
 	}
+
+	/// Returns `true` if we already took all the necessary samples.
+	pub fn is_complete(&self) -> bool {
```
Comment (AndreiEres) [inline polkadot/node/network/protocol/src/grid_topology.rs:546]: nit: we could use it at line 531

**Review by AndreiEres — COMMENTED** [review]
Great job!

**Review by AndreiEres — APPROVED** [review]

**Comment by alexggh** [comment]
Double checked that things are working as expected on versi as well. Merging it now.

**Pull request merged**